p2panda-rs/src/graph/graph.rs

```diff
@@ -0,0 +1,495 @@
+            !node.previous().is_empty() || !node.next().is_empty()
+        } else {
+            false
+        }
+    }
+
+    /// Returns the keys for nodes which follows this node key.
+    pub fn get_next(&'a self, key: &str) -> Option<Vec<String>> {
+        self.get_node(key).map(|node| node.next())
+    }
+
+    /// Returns the keys for nodes which precede this node key.
+    pub fn get_previous(&'a self, key: &str) -> Option<Vec<String>> {
+        self.get_node(key).map(|node| node.previous())
+    }
+
+    /// Returns true if this node key is a merge node.
+    pub fn is_merge_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_merge(),
+            None => false,
+        }
+    }
+
+    /// Returns true if this node key is a branch node.
+    pub fn is_branch_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_branch(),
+            None => false,
+        }
+    }
+
+    /// Returns true if this node key is a graph tip.
+    pub fn is_tip_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_tip(),
+            None => false,
+        }
+    }
+
+    // NOT IMPLEMENTED //
+    // pub fn invalidate_keys(&self, keys: Vec<String>) {}
+
+    /// Returns a reference to the root node of this graph.
+    pub fn root_node(&self) -> &Node<T> {
+        self.0.values().find(|node| node.is_root()).unwrap()
+    }
+
+    /// Returns the root node key.
+    pub fn root_node_key(&self) -> String {
+        self.0.values().find(|node| node.is_root()).unwrap().key()
+    }
+
+    /// Check if all a nodes dependencies have been visited.
+    fn dependencies_visited(&self, sorted: &[&Node<T>], node: &Node<T>) -> bool {
+        let mut has_dependencies = true;
+        let previous_nodes = node.previous();
+
+        for node_key in previous_nodes {
+            let node = self.get_node(&node_key).unwrap();
+            if !sorted.contains(&node) {
+                has_dependencies = false
+            }
+        }
+
+        has_dependencies
+    }
+
+    /// Returns the next un-visited node following the passed node.
+    fn next(&'a self, sorted: &[&Node<T>], node: &Node<T>) -> Option<Vec<&'a Node<T>>> {
+        let mut next_nodes: Vec<&'a Node<T>> = Vec::new();
+
+        for node_key in node.next() {
+            let node = self.get_node(&node_key).unwrap();
+            if !sorted.contains(&node) {
+                next_nodes.push(node)
+            }
+        }
+
+        if next_nodes.is_empty() {
+            return None;
+        };
+        next_nodes.sort_by_key(|node_a| node_a.key());
+        next_nodes.reverse();
+        Some(next_nodes)
+    }
+
+    /// Sorts the graph topologically and returns the sorted
+    pub fn walk_from(&'a self, key: &str) -> Result<GraphData<T>, GraphError> {
+        let root_node = self.get_node(key).unwrap();
+        let mut queue = vec![root_node];
```

Comment:
It would be very helpful to have a short description of what each of these variables defined at the top mean.